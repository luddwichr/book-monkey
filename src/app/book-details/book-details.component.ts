import {Component, OnInit} from '@angular/core';
import {ActivatedRoute, Router} from '@angular/router';
import {flatMap} from 'rxjs/operators';
import {Book} from '../shared/book';
import {BookStoreService} from '../shared/book-store.service';

const isbnParameter = 'isbn';
@Component({
  selector: 'bm-book-details',
  templateUrl: './book-details.component.html',
  styles: []
})
export class BookDetailsComponent implements OnInit {
  book: Book;

  constructor(private bookStoreService: BookStoreService, private route: ActivatedRoute, private router: Router) {}

  ngOnInit(): void {
    this.route.paramMap
      .pipe(flatMap(params => this.bookStoreService.getByIsbn(params.get(isbnParameter))))
      .subscribe(book => {
        this.book = book;
      });
  }

  getRating(num: number) {
    return new Array(num);
  }

  removeBook() {
    this.bookStoreService.remove(this.book).subscribe(() => this.router.navigate(['../'], {relativeTo: this.route}));
  }
}
