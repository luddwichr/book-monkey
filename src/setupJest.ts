import 'jest-preset-angular';
import {Mock} from 'ts-mockery';

Mock.configure('jest');
