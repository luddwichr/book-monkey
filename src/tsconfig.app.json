{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "types": []
  },
  "exclude": ["testing", "setupJest.ts", "**/*.spec.ts"]
}
