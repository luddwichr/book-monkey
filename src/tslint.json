{
  "extends": "../tslint.json",
  "rules": {
    "directive-selector": [true, "attribute", "bm", "camelCase"],
    "component-selector": [true, ["element", "attribute"], "bm", "kebab-case"]
  }
}
