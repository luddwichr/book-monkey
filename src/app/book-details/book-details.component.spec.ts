import {async, ComponentFixture, fakeAsync, TestBed, tick} from '@angular/core/testing';
import {By} from '@angular/platform-browser';
import {ActivatedRoute, Router} from '@angular/router';
import {ActivatedRouteStub, asyncData, BookBuilder} from 'src/testing';
import {anyString, deepEqual, instance, mock, verify, when} from 'ts-mockito';
import {BookStoreService} from '../shared/book-store.service';
import {BookDetailsComponent} from './book-details.component';

describe('BookDetailsComponent', () => {
  const bookStoreServiceMock: BookStoreService = mock(BookStoreService);
  const routerMock: Router = mock(Router);
  const activeRouteStub = new ActivatedRouteStub({isbn: '123'});

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      providers: [
        {provide: BookStoreService, useValue: instance(bookStoreServiceMock)},
        {provide: ActivatedRoute, useValue: activeRouteStub},
        {provide: Router, useValue: instance(routerMock)}
      ],
      declarations: [BookDetailsComponent]
    }).compileComponents();
  }));

  function setup() {
    const fixture: ComponentFixture<BookDetailsComponent> = TestBed.createComponent(BookDetailsComponent);
    const component: BookDetailsComponent = fixture.componentInstance;
    return {fixture, component};
  }

  it('should render the details for a book', fakeAsync(() => {
    const {fixture, component} = setup();
    const book = {
      isbn: '123',
      title: 'someTitle',
      authors: ['A', 'B'],
      published: new Date(2019, 4, 30),
      rating: 3,
      thumbnails: [{url: 'someUrl', title: 'someImageTitle'}],
      subtitle: 'someSubTitle',
      description: 'someDescription'
    };
    when(bookStoreServiceMock.getByIsbn(book.isbn)).thenReturn(asyncData(book));

    fixture.detectChanges();
    tick();
    fixture.detectChanges();

    expect(component.book).toBe(book);
    expect(fixture.nativeElement).toMatchSnapshot();
  }));

  it('should not render anything if no book is given', fakeAsync(() => {
    const {fixture, component} = setup();
    when(bookStoreServiceMock.getByIsbn(anyString())).thenReturn(asyncData(null));

    fixture.detectChanges();
    tick();
    fixture.detectChanges();

    expect(component.book).toBeNull();
    expect(fixture.nativeElement.textContent).toBeFalsy();
  }));

  it('should request to delete the book if the delete button is clicked', fakeAsync(() => {
    const {fixture} = setup();
    const book = BookBuilder.createSomeBookWithIsbn('123');
    when(bookStoreServiceMock.getByIsbn(book.isbn)).thenReturn(asyncData(book));
    when(bookStoreServiceMock.remove(book)).thenReturn(asyncData(null));

    fixture.detectChanges();
    tick();
    fixture.detectChanges();

    fixture.debugElement.query(By.css('#delete-button')).nativeElement.click();
    tick();

    verify(routerMock.navigate(deepEqual(['../']), deepEqual({relativeTo: activeRouteStub}))).called();
  }));
});
