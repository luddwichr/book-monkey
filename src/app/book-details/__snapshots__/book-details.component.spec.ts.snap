// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetailsComponent should render the details for a book 1`] = `
<div
  id="root0"
  ng-version="8.0.0"
>
  
  <div
    class="ui grid"
  >
    <div
      class="sixteen wide column right aligned"
    >
      <div
        class="ui tiny red labeled icon button"
        id="delete-button"
      >
        <i
          class="remove icon"
        />
        Buch l√∂schen 
      </div>
    </div>
    <div
      class="four wide column"
    >
      <img
        alt="someImageTitle"
        class="ui large image"
        src="someUrl"
      />
    </div>
    <div
      class="twelve wide column"
    >
      <h1
        class="ui header"
      >
        someTitle
      </h1>
      <h3
        class="ui header"
      >
        someSubTitle
      </h3>
      <div
        class="ui divider"
      />
      <div
        class="ui grid"
      >
        <div
          class="four wide column"
        >
          <h4
            class="ui header"
          >
            AutorInnen
          </h4>
          
          <span>
             A
            
            <span>
              , 
            </span>
          </span>
          <span>
             B
            
          </span>
        </div>
        <div
          class="four wide column"
        >
          <h4
            class="ui header"
          >
            ISBN
          </h4>
           123 
        </div>
        <div
          class="four wide column"
        >
          <h4
            class="ui header"
          >
            Erschienen
          </h4>
           May 30, 2019 
        </div>
        <div
          class="four wide column"
        >
          <h4
            class="ui header"
          >
            Rating
          </h4>
          
          <i
            class="yellow star icon"
          />
          <i
            class="yellow star icon"
          />
          <i
            class="yellow star icon"
          />
        </div>
      </div>
    </div>
    <div
      class="sixteen wide column"
    >
      <h3
        class="ui header"
      >
        Beschreibung
      </h3>
       someDescription 
    </div>
    <div
      class="sixteen wide column"
    >
      <div
        class="ui small images"
      >
        
        <img
          alt="someImageTitle"
          src="someUrl"
        />
      </div>
    </div>
  </div>
</div>
`;
